### ISSUE TITLE
BigQuery Row Level Security Policy

### ISSUE THREAD

**Issue by aaronclong** [hashicorp/terraform-provider-google#17464]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to a user, that user is claiming responsibility for the issue.
* Customers working with a Google Technical Account Manager or Customer Engineer can ask them to [reach out internally](https://github.com/hashicorp/terraform-provider-google/wiki/Customer-Contact#raising-gcp-internal-issues-with-the-provider-development-team) to expedite investigation and resolution of this issue.

### Description

I can't seem to find any documentation or resources for [BigQuery's Row Level Security](https://cloud.google.com/bigquery/docs/managing-row-level-security#create-policy).  Do existing resources provide for this already, or does a new resource or modification of existing resource need to be crafted?

The only workaround to this seems to be a to [create BigQuery Job to add the permissions](https://stackoverflow.com/a/77111370).

### New or Affected Resource(s)

* google_bigquery_table
* google_bigquery_table_iam

### Potential Terraform Configuration

```tf 
```

### References

https://cloud.google.com/bigquery/docs/managing-row-level-security#create-policy

b/328133777

**Comment by melinath** [comment]
Note from triage: This doesn't seem to be supported at the REST API layer, only by calls to BigQuery directly, which means we can't support it in the provider at this time. Support would need to be added to the REST API first.

**Comment by aaronclong** [comment]
> Note from triage: This doesn't seem to be supported at the REST API layer, only by calls to BigQuery directly, which means we can't support it in the provider at this time. Support would need to be added to the REST API first.

That's good know, thank you for pointing that out. I wonder if I can make feature request to Google and link that here.

**Comment by wj-chen** [comment]
RowAccessPolicy is its own GCP resource (https://cloud.google.com/bigquery/docs/reference/rest/v2/rowAccessPolicies), so it'll likely translate to a new resource in the Terraform provider. As mentioned, there is no official API to manage RowAccessPolicy at the time. The standard procedure would be to make the API available first before adding it to the provider. Since the it's possible to manage it using DDL statements, it may be possible to implement the new Terraform resource in such a way that behind the scenes we just execute the DDL statement instead of calling an API, but it's generally not ideal as e.g. it requires careful surface design to be as forward-compatible as possible to match the potential API surface when an API does become available, and it requires the user to have IAM permission for e.g. `jobs.create` for now to run the BigQuery Job but the actual `rowAccessPolicies.create` permission in the future, and the transition may break users. If using the Job resource separately unblocks you, I'd suggest doing that for now.

**Comment by wj-chen** [comment]
There is now work underway to support managing Row Level Security Policies through the API which will unblock adding support for this resource in Terraform. A tentative timeline for availability is Q4. Please continue to follow this bug for updates.

**Comment by aaronclong** [comment]
> https://issuetracker.google.com/issues/229861829

@wj-chen is there any source that you can link here?

**Comment by wj-chen** [comment]
This work is scheduled for Q4. Progress will be communicated here in this issue.

**Comment by wj-chen** [comment]
The underlying API support is currently Google-internal and waiting for a public release in early January, after all the waves of end-of-year deployment freeze. The Terraform side of development can begin once the underlying API is ready, tentatively Q1 2025.

**Comment by itaispiegel** [comment]
We're hoping to see this feature very soon 🙏🏻

**Comment by wj-chen** [comment]
We are experiencing a delay in the production rollouts of the underlying API and client library that our Terraform provider depends on for this feature. The new plan is to start the Terraform work in Q2 2025 after the dependencies are available. Please continue to follow this issue for updates, and thank you for your patience.

**Comment by wj-chen** [comment]
Thanks all for your patience. https://github.com/GoogleCloudPlatform/magic-modules/pull/13932 is in review and will add a new resource `google_bigquery_row_access_policy`.
